model/textparse/protobufparse.go

```diff
@@ -418,106 +395,138 @@ func (p *ProtobufParser) CreatedTimestamp() *int64 {
 				if !strings.HasSuffix(name[:len(name)-6], unit) || len(name)-6 < len(unit)+1 || name[len(name)-6-len(unit)-1] != '_' {
 					return EntryInvalid, fmt.Errorf("unit %q not a suffix of counter %q", unit, name)
 				}
 			} else if !strings.HasSuffix(name, unit) || len(name) < len(unit)+1 || name[len(name)-len(unit)-1] != '_' {
 				return EntryInvalid, fmt.Errorf("unit %q not a suffix of metric %q", unit, name)
 			}
 		}
-		p.metricBytes.Reset()
-		p.metricBytes.WriteString(name)
-
+		p.seriesBytes.Reset()
+		p.seriesBytes.WriteString(name)
 		p.state = EntryHelp
 	case EntryHelp:
-		if p.mf.Unit != "" {
+		if p.dec.Unit != "" {
 			p.state = EntryUnit
 		} else {
 			p.state = EntryType
 		}
 	case EntryUnit:
 		p.state = EntryType
 	case EntryType:
-		t := p.mf.GetType()
-		if (t == dto.MetricType_HISTOGRAM || t == dto.MetricType_GAUGE_HISTOGRAM) &&
-			isNativeHistogram(p.mf.GetMetric()[0].GetHistogram()) {
+		t := p.dec.GetType()
+		if (t == clientpb.MetricType_HISTOGRAM || t == clientpb.MetricType_GAUGE_HISTOGRAM) &&
+			isNativeHistogram(p.dec.GetHistogram()) {
 			p.state = EntryHistogram
 		} else {
 			p.state = EntrySeries
 		}
-		if err := p.updateMetricBytes(); err != nil {
+		if err := p.onSeriesOrHistogramUpdate(); err != nil {
 			return EntryInvalid, err
 		}
-	case EntryHistogram, EntrySeries:
-		if p.redoClassic {
-			p.redoClassic = false
-			p.state = EntrySeries
-			p.fieldPos = -3
-			p.fieldsDone = false
-		}
-		t := p.mf.GetType()
-		if p.state == EntrySeries && !p.fieldsDone &&
-			(t == dto.MetricType_SUMMARY ||
-				t == dto.MetricType_HISTOGRAM ||
-				t == dto.MetricType_GAUGE_HISTOGRAM) {
-			p.fieldPos++
-		} else {
-			p.metricPos++
+	case EntrySeries:
+		// Potentially a second series in the metric family.
+		t := p.dec.GetType()
+		if t == clientpb.MetricType_SUMMARY ||
+			t == clientpb.MetricType_HISTOGRAM ||
+			t == clientpb.MetricType_GAUGE_HISTOGRAM {
+			// Non-trivial series (complex metrics, with magic suffixes).
+
+			// Did we iterate over all the classic representations fields?
+			// NOTE: p.fieldsDone is updated on p.onSeriesOrHistogramUpdate.
+			if !p.fieldsDone {
+				// Still some fields to iterate over.
+				p.fieldPos++
+				if err := p.onSeriesOrHistogramUpdate(); err != nil {
+					return EntryInvalid, err
+				}
+				return p.state, nil
+			}
+
+			// Reset histogram fields.
 			p.fieldPos = -2
 			p.fieldsDone = false
 			p.exemplarPos = 0
+
 			// If this is a metric family containing native
-			// histograms, we have to switch back to native
-			// histograms after parsing a classic histogram.
-			if p.state == EntrySeries &&
-				(t == dto.MetricType_HISTOGRAM || t == dto.MetricType_GAUGE_HISTOGRAM) &&
-				isNativeHistogram(p.mf.GetMetric()[0].GetHistogram()) {
+			// histograms, it means we are here thanks to redoClassic state.
+			// Return to native histograms for the consistent flow.
+			if t == clientpb.MetricType_HISTOGRAM || t == clientpb.MetricType_GAUGE_HISTOGRAM &&
```

Comment:
Missing paren makes histogram parsing fail:

```suggestion
			if (t == clientpb.MetricType_HISTOGRAM || t == clientpb.MetricType_GAUGE_HISTOGRAM) &&
```